Fixes #15630: Add file content type API

ehelms opened the pull request.

ehelms commented:
Will require https://github.com/Katello/katello/pull/5770

daviddavis commented on app/models/katello/file.rb:
```diff
@@ -0,0 +1,32 @@
+
+    def update_from_json(json)
+      keys = %w(name checksum)
+      custom_json = json.clone.delete_if { |key, _value| !keys.include?(key) }
```
Seems weird to be using `clone` and `delete_if` instead of just `reject`. Better yet, just use slice:

```ruby
custom_json = json.slice('name', 'checksum')
```

ehelms replied:
Sorry I missed this -- do you think this changed should be made everywhere? Given every repository type implements this function?

daviddavis replied:
I think just here is fine. Unless you feel the need to update it everywhere. Conversely, if you want to keep this the way it is to be consistent with other repository types, that's fine too. Your call.

daviddavis replied:
[Bam](https://github.com/Katello/katello/pull/6279)

ehelms commented:
Rebased on top of the recently merged file PR

ehelms commented:
This still needs some tests added to it but if someone wants to review it for the basic design and data model that would be greatly appreciated.

jlsherrill commented:
getting this when syncing:

```
uninitialized constant Actions::Pulp::Repository::Presenters::FileUnitPresenter (NameError)
/home/vagrant/git/katello/app/lib/actions/pulp/repository/sync.rb:89:in `presenter'
/home/vagrant/git/katello/app/lib/actions/helpers/presenter.rb:9:in `humanized_output'
/home/vagrant/git/katello/app/lib/
```

jlsherrill commented:
Error when indexing content

```
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column "file_unit_id" does not exist
LINE 1: SELECT file_unit_id FROM "katello_repository_files" WHERE "k...
               ^
: SELECT file_unit_id FROM "katello_repository_files" WHERE "katello_repository_files"."repository_id" = 141
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.6/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `async_exec'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.6/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `block in exec_no_cache'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.6/lib/active_record/connection_adapters/abstract_adapter.rb:472:in `block in log'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.6/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.6/lib/active_record/connection_adapters/abstract_adapter.rb:466:in `log'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.6/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `exec_no_cache'
	from /home/vagrant/.rvm/gems/ruby-2.2.4/gem
```

ehelms commented:
@jlsherrill updated

daviddavis commented:
What about this comment: https://github.com/Katello/katello/pull/6176#discussion_r70465518

ehelms commented:
@daviddavis addressed

jlsherrill commented on db/migrate/20160808002834_add_files.rb:
```diff
@@ -0,0 +1,29 @@
+      t.timestamps
+      t.string "uuid", :null => false, :limit => 255
+      t.string 'name', :limit => 255
+      t.string 'checksum', :limit => 255
```
Does name and checksum really need to be limited to 255?

ehelms replied:
I was following the other content types that were set to this based on (https://github.com/Katello/katello/pull/5837). If you think I should open this up to unlimited then I shall.

ehelms replied:
@jlsherrill locally I have it looking like this:

```
{
id: 6,
name: "katello-ca-consumer-latest.noarch.rpm",
path: "pops/hey/WHAT",
uuid: "9d03acff-e674-4f49-8753-f900e7d22d12",
checksum: "24065bf79b81abd806fbe989edcfffcc1e3b4bb8a02726cbfbc957de3df166fd"
},
```

Do you think the full path would be better there? or just let users and interfaces construct it?

jlsherrill replied:
I might would just have the full path and not the partial path.  That seems simpler and the user can do with it what they want.
```
name: "katello-ca-consumer-latest.noarch.rpm",
path: "pops/hey/WHAT/katello-ca-consumer-latest.noarch.rpm"
```

ehelms commented:
Updated:
 * renamed controller and view paths to 'file_units' to match with model name
 * add `path` to model and storing the full path or what Pulp considers the 'name'
 * name is now just the basename of the file
 * added path as a scoped search option and to the API response

jlsherrill commented:
https://github.com/Katello/katello/blob/master/app/controllers/katello/concerns/api/v2/repository_content_controller.rb#L187-L204  needs to be updated otherwise requests for nonexistant files throws an error:

GET http://robot:3000/katello/api/v2/files/zoom

"Can't find resource class: Katello::FileUnit"

ehelms commented:
@jlsherrill Good catch -- updated

jlsherrill commented:
APJ

ehelms commented:
[test]
